File: agent/skills/skill_registry.py
from __future__ import annotations

import logging

from skills.skill import Skill

logger = logging.getLogger(__name__)


class SkillRegistry:
    """Registry that holds all available skills and supports dynamic registration."""

    def __init__(self) -> None:
        self._skills: dict[str, Skill] = {}

    def register(self, skill: Skill) -> None:
        if skill.name in self._skills:
            logger.warning("Overwriting existing skill: %s", skill.name)
        self._skills[skill.name] = skill

    def get(self, name: str) -> Skill | None:
        return self._skills.get(name)

    def list_all(self) -> list[Skill]:
        return sorted(self._skills.values(), key=lambda s: s.priority, reverse=True)

    def catalog_prompt(self) -> str:
        """Build a prompt snippet listing all available skills for LLM selection."""
        lines = ["Available skills:"]
        for skill in self.list_all():
            lines.append(f"- {skill.name}: {skill.description}")
        return "\n".join(lines)

    def brief_prompt(self, names: list[str]) -> str:
        """Build a prompt from the brief of selected skills (progressive disclosure)."""
        parts = []
        for name in names:
            skill = self.get(name)
            if skill is not None:
                parts.append(f"[{skill.name}] {skill.brief}")
        return "\n".join(parts)

    def expand_skill(self, name: str) -> str:
        """Return the full system_prompt for a single skill (on-demand expansion)."""
        skill = self.get(name)
        if skill is None:
            return ""
        return skill.system_prompt

File: agent-ts/src/skills/core/SkillRegistry.ts
import type { Skill } from "../model/Skill.js";

export class SkillRegistry {
  private readonly skills = new Map<string, Skill>();

  register(skill: Skill): void {
    if (this.skills.has(skill.name)) {
      // 覆盖同名 skill 时保持 Python 版的可观察行为。
      console.warn(`Overwriting existing skill: ${skill.name}`);
    }
    this.skills.set(skill.name, skill);
  }

  get(name: string): Skill | undefined {
    return this.skills.get(name);
  }

  listAll(): Skill[] {
    return [...this.skills.values()].sort((a, b) => b.priority - a.priority);
  }

  catalogPrompt(): string {
    const lines = ["Available skills:"];
    for (const skill of this.listAll()) {
      lines.push(`- ${skill.name}: ${skill.description}`);
    }
    return lines.join("\n");
  }

  briefPrompt(names: readonly string[]): string {
    const parts: string[] = [];
    for (const name of names) {
      const skill = this.get(name);
      if (skill) {
        parts.push(skill.brief);
      }
    }
    return parts.join("\n");
  }

  expandSkill(name: string): string {
    return this.get(name)?.systemPrompt ?? "";
  }
}
